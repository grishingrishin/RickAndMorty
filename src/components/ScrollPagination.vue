<script setup>
import { defineEmits, defineProps, onMounted, onUnmounted } from 'vue';

import { throttle } from '@/utils/helpers';

const GLOBAL_EVENTS = [
	{
		name: 'scroll',
		handler: throttle(onScroll),
	},
	{
		name: 'resize',
		handler: throttle(onScroll),
	},
	{
		name: 'wheel',
		handler: throttle(onWheel),
	},
];

const THRESHOLD_PART = 4;

const props = defineProps({
	disabled: {
		type: Boolean,
		default: false,
	},
});

const emit = defineEmits(['update']);

onMounted(addEventListeners);
onUnmounted(removeEventListeners);

function addEventListeners() {
	if (props.disabled) {
		return;
	}

	GLOBAL_EVENTS.forEach(evt => {
		window.addEventListener(evt.name, evt.handler);
	});
}

function removeEventListeners() {
	if (props.disabled) {
		return;
	}

	GLOBAL_EVENTS.forEach(evt => {
		window.removeEventListener(evt.name, evt.handler);
	});
}

function onScroll() {
	const scrollHeight = document.documentElement.scrollHeight;
	const screenHeight = window.innerHeight;
	const scrolled = window.scrollY;

	const threshold = scrollHeight - screenHeight / THRESHOLD_PART;
	const position = scrolled + screenHeight;

	if (position >= threshold) {
		emit('update');
	}
}

function onWheel() {
	const scrollHeight = document.documentElement.scrollHeight;
	const screenHeight = window.innerHeight;
	const scrolled = window.scrollY;

	const threshold = scrollHeight - screenHeight;

	if (scrolled >= threshold) {
		emit('update');
	}
}
</script>

<template>
	<slot />
</template>
