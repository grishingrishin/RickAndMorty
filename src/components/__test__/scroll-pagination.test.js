import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import ScrollPagination from '@/components/ScrollPagination.vue';

describe('ScrollPagination', () => {
	test('Mounted pure component', () => {
		expect(ScrollPagination).toBeTruthy();

		const wrapper = mount(ScrollPagination);

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Emitting an update event', () => {
		expect(ScrollPagination).toBeTruthy();

		const wrapper = mount(ScrollPagination);

		genEvent('scroll');

		expect(wrapper.emitted()).toHaveProperty('update');
	});

	test('Disabled emitting an update event', () => {
		expect(ScrollPagination).toBeTruthy();

		const wrapper = mount(ScrollPagination, {
			props: {
				disabled: true,
			},
		});

		genEvent('scroll');

		expect(wrapper.emitted()).not.toHaveProperty('update');
	});

	function genEvent(evtName, params) {
		const customEvent = new CustomEvent(evtName, params);
		window.dispatchEvent(customEvent);
	}
});
