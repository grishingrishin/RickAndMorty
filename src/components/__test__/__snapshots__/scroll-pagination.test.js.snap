// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScrollPagination > Mounted pure component 1`] = `""`;
